api/users.go

```diff
@@ -53,35 +60,383 @@ func (self *ApiServer) SetPassword(
+		}
+		if !createUser {
+			removed := map[string]*api_proto.Org{}
+			for _, org := range user_record.Orgs {
+				removed[org.Id] = org
+
+			}
+			for _, org := range in.User.Orgs {
+				delete(removed, org.Id)
+			}
+			required_orgs := make([]*api_proto.Org, 0, len(removed))
+			idx := 0
+			for _, v := range removed {
+				required_orgs[idx] = v
+				idx += 1
+			}
+
+			if !checkOrgMembership(required_orgs, active_user.Orgs) {
+				return nil, status.Error(codes.PermissionDenied,
+					"User is not allowed to remove users from orgs without the org admin role or being a member of the org.")
+			}
+		}
+	}
+
+	org_manager, err := services.GetOrgManager()
+	if err != nil {
+		return nil, err
+	}
+
+	config_obj, err := org_manager.GetOrgConfig(services.ROOT_ORG_ID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = setPassword(config_obj, in.User, in.Password, createUser)
+	if err != nil {
+		return nil, err
+	}
+
+	err = grantUserRoles(config_obj, &org_manager, in.User)
+	if err != nil {
+		return nil, err
+	}
+
+	user_record.Email = in.User.Email
+	user_record.Picture = in.User.Picture
+	user_record.VerifiedEmail = in.User.VerifiedEmail
+
+	// This must go after setting the password since SetPassword clears the Locked flag
+	user_record.Locked = in.User.Locked
+
+	err = users_manager.SetUser(ctx, user_record)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := "users: Updating user record via API"
+	if createUser {
+		msg = "users: Creating user record via API"
+	}
+
+	logger := logging.GetLogger(org_config_obj, &logging.Audit)
+	logger.WithFields(logrus.Fields{
+		"Username":  active_user.Name,
+		"Principal": user_record.Name,
+	}).Info(msg)
+
+	return &emptypb.Empty{}, nil
+}
+
+func (self *ApiServer) CreateUser(ctx context.Context,
+				  in *api_proto.ChangeUserRequest) (*emptypb.Empty, error) {
+	return self.ChangeUser(ctx, in, true)
+}
+
+func (self *ApiServer) UpdateUser(ctx context.Context,
+				  in *api_proto.ChangeUserRequest) (*emptypb.Empty, error) {
+	return self.ChangeUser(ctx, in, false)
+}
+
+func (self *ApiServer) DeleteUser(
+	ctx context.Context,
+	in *api_proto.UserRequest) (*emptypb.Empty, error) {
+
+	users_manager := services.GetUserManager()
+	active_user, org_config_obj, err := users_manager.GetUserFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	permissions := acls.SERVER_ADMIN
+	perm, err := acls.CheckAccess(org_config_obj, active_user.Name, permissions)
+	if !perm || err != nil {
+		return nil, status.Error(codes.PermissionDenied,
+			"User is not allowed to delete users.")
+	}
+
+	user_record, err := users_manager.GetUserWithHashes(ctx, in.Name)
+	if err == services.UserNotFoundError {
+		return nil, status.Errorf(codes.NotFound, "User %s does not exist.", in.Name)
+	}
+
+	org_admin, err := acls.CheckAccess(org_config_obj, active_user.Name, acls.ORG_ADMIN)
+	if err != nil {
+		return nil, err
+	}
+
+	if !org_admin && checkOrgMembership(user_record.Orgs, active_user.Orgs) == false {
```

Comment:
You generally dont need to check org membership because the ACL exists in each org anyway so the check above ensures the user is an admin in the org.

Reply:
I think this ends up being a fundamental issue with this entire PR. I wrote most of this against 0.6.4 before the organizations were introduced and had to update it against 0.6.6 (and now 0.6.7-rc1) to submit it. The user admin UI is not taking into account the current organization of the principal. It's displaying and operating in a global view, but constraining it by the list of orgs the principal is a member of and, when appropriate, an intersection of the orgs of the user being updated/deleted. I was basing that on  `VelociraptorUser`  having a list of orgs and a set of roles and mistakenly assuming that meant that the roles applied to every Org, when it probably means that the roles correspond to the current org.

I'll need to go back and rework this.

Reply:
All you really need to do is to delegate to the VQL functions to do the actual work - the API should be a thin wrapper around the VQL which is supposed to do the right thing in all cases

Reply:
I suppose I don't have a clear grasp on what the roles mean in the scope of orgs and where the separation of the org takes effect.

I see that ORG_ADMIN in the current org allows:
- COLLECT_CLIENT permission in _any_ org via `v1/CreateHunt` and `hunt_add()`
- Listing users across orgs
- Creating new orgs

SERVER_ADMIN is typically evaluated against the current org, but often the operations the SERVER_ADMIN role can perform aren't limited to that org. For example, `user_grant()` checks for SERVER_ADMIN in the current org but then allows the caller to operate on roles in any org. `user_grant()` can also be used to modify the ACLs in each org separately so that having an ACL in one org doesn't mean it exists in other.

I suppose my code was trying to refine this a bit, at least for the API, so that SERVER_ADMIN only applies to the orgs to which the user belongs. ORG_ADMIN allows SERVER_ADMIN to be global, but the the direct implication of SERVER_ADMIN -> ORG_ADMIN means that distinction doesn't exist.

Do orgs only exist for the lesser privileged roles and SERVER_ADMIN is intended to be global? If that's the case, we should probably avoid inconsistencies with SERVER_ADMIN only being set on some orgs.

Reply:
ORG_ADMIN always allows full org control. I guess it is an oversight that a SERVER_ADMIN can grant ORG_ADMIN to a user without having ORG_ADMIN themselves. 

This new permission model is pretty new so we do need to think through some of these implications. This is why it is important to have consistent calls into the same well tested set of functions instead of reimplement the logic in slightly different ways in different places.

If you dont mind I can help with the PR by refactoring the VQL functions to make them easier to call from the API functions. Otherwise this is really a great addition !  Thanks for kicking the tires on the permission system too, clearly we have some way to go on it :-).

The whole point of the multi org feature is to make it safe for an org admin to provision an org and give full admin rights (SERVER_ADMIN) to a user in that org without escalation to other orgs. One potential escalation path is arbitrary file write, execve and other dangerous plugins which #2214 is enforcing so we are making progress towards that goal - but we are not there yet!

Reply:
I've already started in on refactoring the VQL stuff for user manipulation.  The general idea is to factor out fetching the ACLManager from VQL so that both the API and VQL callers can pass it in to a common implementation under services/users. I'm using a request struct for the refactor that gets replaced by the api_proto version later, so that there's no real difference on the VQL side and a straight line on the API side. In doing so I found a bit of a weakness in my original approach: you have to essentially specify the entire user with all the fields every time and when the gRPC http proxy ends up removing booleans if they're false or unset, they also always need to be specified. Not ideal. So now I'm using two maps of string->string and string->bool to indicate new values for fields.